README.md
# nweb
a tiny c implementation of a http server

nweb stands for Nigel's Web server but I some times call it a nano web server (i.e. very small).

## how to use
`./nweb [port] [root directory]`
`./nweb 8181 /home/nwebdir`

## how to compile
`gcc -O -DLINUX nweb.c -o nweb`

-----
this is part of [the nmon project](https://sourceforge.net/projects/nmon) from Nigel Griffiths


changelog.md
* nweb.c -- a tiny c implementation of a http server
nweb stands for Nigel's Web server but I some times call it a nano web server (i.e. very small).


# changelog
This file contains lots of hints & tips on compiling and running the nweb code

## version 24
* Website move due to it be oddly dropped with no warning
* Due to feedback `chdir()` was changed to `chroot()`
  This massively increases security as clever URL trickiery can never get a non-home directory file sent to the browser.
* Please note this code is to illustrate the algorythm of a webserver in a minimum number of lines of code - it is not ment to be absolutely perfect C code. 


## version 23

* Bug fixed: was duplicating errors in the nweb.log file. thanks to Kieran Grant for stopping this and pointing out the fix.

* Added support for favicon.ico if nothing elase this will stop annoying errors in the log file.  Most browsers on first encountering a webpage also ask for this file.  The name mean favourite icon - Wikipedia has more on this.  It is a tiny Bit Map image (normally called .bmp and
BMP editors can be used to create one - I used Windows Paint)
I uses a very simple 16x16 pixels and 256 colour to keep the size and complexity down.
To add support I added .ico file extension to the allowed extensions data structure and the sample favicon.ico file.
This is then displyed by the browser at the little graphic next to the URL. The one I made looks like this withe blue background.
```
+---+
|n  |
|Web|
+---+
```

## version 22

This nweb is a web server that will respond to simple web browser 
requests for static files.
This version is exactly 200 lines long. 
I started with a target 100 lines and it worked fine too but then 
added comments, file type checks,
security checks, sensible directory checks and logging.

To compile this you need a basic C compiler use: cc nweb22.c -o nweb
If you want it to run faster you could include optimisation: 
cc -O2 nweb22.c -o nweb

Then to run it as root for the test files:
1) Place the index.html file and nigel.jpg in to a sensible directory 
	Note: not any system directories like / or /tmp 
	A director in /home would be a good idea - below we use /home/nigel/web

2) Make these files readable
	chmod ugo+r /home/nigel/web

3) Decide the port number
	Port number 80 is the web server default and assumed by web browsers.
	you might have to be the root user to use that  port number.
	If the machine is already running a web server then it will 
	probably be using port 80 so you can't us 80.
	I test using port 8181 but you have to make up your mind.

4) Start the nweb server:  /home/nigel/bin/nweb 80   /home/nigel/web
	or if using 8181   /home/nigel/bin/nweb 8181 /home/nigel/web

	If the program and files are all in /home/nigel/all
	You could use:
		cd /home/nigel/all
		rm -f nweb.log
		./nweb 80 .
		ps -ef | grep nweb 
		tail -f nweb.log

	See below for explanations for the rm, ps and tail

5) Note that any suitable files in the directory 2nd argument above 
could be served by the nmon web server - so don't have anything secret 
in that directory!

6) Use a browser to test it if using port 80 on a machine hostname 
of abc123.com browser to:
	http://abc123.com/index.html

   If using port 8181
	http://abc123.com:8181/index.html

7) Running nweb as a regualr user
You can run nweb as a regular user on some operating systems but not all. I find using sudo works on some: 	`sudo /home/nigel/bin/nweb 80   /home/nigel/web`. If you try and the log file reports socect connections errors worth 
	trying the 8181 port or higher numbers.  This is because 
	lower port numbers are reserved for root user use.

8) VERY IMPORTANT
nweb will behave like a daemon process, so it disconnects from the 
users command shell, goes into the back ground, closes input and 
output I/O & protects itself from you logging off.  
It will look like it stopped, when it is in fact still running in 
the background - check with `ps -ef | grep nweb`
Also note you will not see errors or warnings messages - they go in the log file. Look in the nweb.log file for problems starting up or browsers connecting and requesting pages with: `cat nweb.log` or: `tail -f nweb.log`. I find it good to remove the nweb.log before I start nweb or it is easy to get confused with previous error messages in the log.

9) As nweb runs as a daemon process it will try to run forever and not conntected to your user or terminal session.  Logging out will not effect it. To stop nweb you have to stop it by a KILL signal. 
Find the nweb process with: `ps -ef | grep nweb`
Then use: `kill -9 PID ` to stop the nweb process.


10) nweb.log Log file - what to look for
Good example

Here is an edited down sample nweb.log file. 
Note the first line = "starting" looks good as there is no following error.

```
 INFO: nweb starting:8181:8913126
 INFO: request:GET /index.html HTTP/1.1 [[1KB of deleted request stuff here]]:2
 INFO: SEND:index.html:2
 INFO: Header:HTTP/1.1 200 OK
Server: nweb/21.0
Content-Length: 239
Connection: close
Content-Type: text/html

 INFO: request:GET /nigel.jpg HTTP/1.1  [[1KB of deleted request stuff here]]:3
 INFO: SEND:nigel.jpg:3
 INFO: Header:HTTP/1.1 200 OK
Server: nweb/21.0
Content-Length: 10184
Connection: close
Content-Type: image/jpg
``` 

Bad example
A failure to start due to the port number is not aallowed or in use looks like:
```
 INFO: nweb starting:80:8323296
ERROR: system call:bind Errno=13 exiting pid=8323296
```


nweb.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <fcntl.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#define VERSION 25
#define BUFSIZE 65536 /* larger = more efficient */
#define ERROR      42
#define LOG        44
#define FORBIDDEN 403
#define NOTFOUND  404
#ifndef SIGCLD
#define SIGCLD SIGCHLD
#endif
struct {
	char *ext;
	char *filetype;
} extensions [] = {
	{"gif", "image/gif" },  {"jpg", "image/jpg" }, {"jpeg","image/jpeg"},
	{"png", "image/png" },  {"ico", "image/ico" },  {"zip", "image/zip" },
	{"gz",  "image/gz"  },  {"tar", "image/tar" },  {"htm", "text/html" },
	{"html","text/html" },  {0,0} };

void logger(int type, char *s1, char *s2, int socket_fd)
{
	int fd ;
	char logbuffer[BUFSIZE*2];

	switch (type) {
	case ERROR: (void)sprintf(logbuffer,"ERROR: %s:%s Errno=%d exiting pid=%d",s1, s2, errno,getpid());
		break;
	case FORBIDDEN:
		(void)write(socket_fd, "HTTP/1.1 403 Forbidden\nContent-Length: 185\nConnection: close\nContent-Type: text/html\n\n<html><head>\n<title>403 Forbidden</title>\n</head><body>\n<h1>Forbidden</h1>\nThe requested URL, file type or operation is not allowed on this simple static file webserver.\n</body></html>\n",271);
		(void)sprintf(logbuffer,"FORBIDDEN: %s:%s",s1, s2);
		break;
	case NOTFOUND:
		(void)write(socket_fd, "HTTP/1.1 404 Not Found\nContent-Length: 136\nConnection: close\nContent-Type: text/html\n\n<html><head>\n<title>404 Not Found</title>\n</head><body>\n<h1>Not Found</h1>\nThe requested URL was not found on this server.\n</body></html>\n",224);
		(void)sprintf(logbuffer,"NOT FOUND: %s:%s",s1, s2);
		break;
	case LOG: (void)sprintf(logbuffer," INFO: %s:%s:%d",s1, s2,socket_fd); break;
	}
	/* No checks here, nothing can be done with a failure anyway */
	if((fd = open("nweb.log", O_CREAT| O_WRONLY | O_APPEND,0644)) >= 0) {
		(void)write(fd,logbuffer,strlen(logbuffer));
		(void)write(fd,"\n",1);
		(void)close(fd);
	}
	if(type == ERROR || type == NOTFOUND || type == FORBIDDEN) exit(3);
}

/* this is a child web server process, so we can exit on errors */
void web(int fd, int hit)
{
	int j, file_fd, buflen;
	long i, ret, len;
	char * fstr;
	static char buffer[BUFSIZE+1]; /* static so zero filled */

	ret =read(fd,buffer,BUFSIZE); 	/* read Web request in one go */
	if(ret == 0 || ret == -1) {	/* read failure stop now */
		logger(FORBIDDEN,"failed to read browser request","",fd);
	}
	if(ret > 0 && ret < BUFSIZE)	/* return code is valid chars */
		buffer[ret]=0;		/* terminate the buffer */
	else buffer[0]=0;
	for(i=0;i<ret;i++)	/* remove CF and LF characters */
		if(buffer[i] == '\r' || buffer[i] == '\n')
			buffer[i]='*';
	logger(LOG,"request",buffer,hit);
	if( strncmp(buffer,"GET ",4) && strncmp(buffer,"get ",4) ) {
		logger(FORBIDDEN,"Only simple GET operation supported",buffer,fd);
	}
	for(i=4;i<BUFSIZE;i++) { /* null terminate after the second space to ignore extra stuff */
		if(buffer[i] == ' ') { /* string is "GET URL " +lots of other stuff */
			buffer[i] = 0;
			break;
		}
	}
	for(j=0;j<i-1;j++) 	/* check for illegal parent directory use .. */
		if(buffer[j] == '.' && buffer[j+1] == '.') {
			logger(FORBIDDEN,"Parent directory (..) path names not supported",buffer,fd);
		}
	if( !strncmp(&buffer[0],"GET /\0",6) || !strncmp(&buffer[0],"get /\0",6) ) /* convert no filename to index file */
		(void)strcpy(buffer,"GET /index.html");

	/* work out the file type and check we support it */
	buflen=strlen(buffer);
	fstr = (char *)0;
	for(i=0;extensions[i].ext != 0;i++) {
		len = strlen(extensions[i].ext);
		if( !strncmp(&buffer[buflen-len], extensions[i].ext, len)) {
			fstr =extensions[i].filetype;
			break;
		}
	}
	if(fstr == 0) logger(FORBIDDEN,"file extension type not supported",buffer,fd);

	if(( file_fd = open(&buffer[5],O_RDONLY)) == -1) {  /* open the file for reading */
		logger(NOTFOUND, "failed to open file",&buffer[5],fd);
	}
	logger(LOG,"SEND",&buffer[5],hit);
	len = (long)lseek(file_fd, (off_t)0, SEEK_END); /* lseek to the file end to find the length */
	      (void)lseek(file_fd, (off_t)0, SEEK_SET); /* lseek back to the file start ready for reading */
          (void)sprintf(buffer,"HTTP/1.1 200 OK\nServer: nweb/%d.0\nContent-Length: %ld\nConnection: close\nContent-Type: %s\n\n", VERSION, len, fstr); /* Header + a blank line */
	logger(LOG,"Header",buffer,hit);
	(void)write(fd,buffer,strlen(buffer));

	/* send file in 8KB block - last block may be smaller */
	while (	(ret = read(file_fd, buffer, BUFSIZE)) > 0 ) {
		(void)write(fd,buffer,ret);
	}
	sleep(1);	/* allow socket to drain before signalling the socket is closed */
	close(fd);
	exit(1);
}

int main(int argc, char **argv)
{
	int i, port, pid, listenfd, socketfd, hit;
	socklen_t length;
	static struct sockaddr_in cli_addr; /* static = initialised to zeros */
	static struct sockaddr_in serv_addr; /* static = initialised to zeros */

	if( argc < 3  || argc > 3 || !strcmp(argv[1], "-?") ) {
		(void)printf("hint: nweb Port-Number Top-Directory\t\tversion %d\n\n"
	"\tnweb is a small and very safe mini web server\n"
	"\tnweb only servers out file/web pages with extensions named below\n"
	"\t and only from the named directory or its sub-directories.\n"
	"\tThere is no fancy features = safe and secure.\n\n"
	"\tExample: nweb 8181 /home/nwebdir &\n\n"
	"\tOnly Supports:", VERSION);
		for(i=0;extensions[i].ext != 0;i++)
			(void)printf(" %s",extensions[i].ext);

		(void)printf("\n\tNot Supported: URLs including \"..\", Java, Javascript, CGI\n"
	"\tNot Supported: directories / /etc /bin /lib /tmp /usr /dev /sbin \n"
	"\tNo warranty given or implied\n\tNigel Griffiths nag@uk.ibm.com\n"  );
		exit(0);
	}
	if( !strncmp(argv[2],"/"   ,2 ) || !strncmp(argv[2],"/etc", 5 ) ||
	    !strncmp(argv[2],"/bin",5 ) || !strncmp(argv[2],"/lib", 5 ) ||
	    !strncmp(argv[2],"/tmp",5 ) || !strncmp(argv[2],"/usr", 5 ) ||
	    !strncmp(argv[2],"/dev",5 ) || !strncmp(argv[2],"/sbin",6) ){
		(void)printf("ERROR: Bad top directory %s, see nweb -?\n",argv[2]);
		exit(3);
	}
	if(chroot(argv[2]) == -1){
		(void)printf("ERROR: Can't Change to directory %s\n",argv[2]);
		exit(4);
	}
	/* Become deamon + unstopable and no zombies children (= no wait()) */
	if(fork() != 0)
		return 0; /* parent returns OK to shell */
	(void)signal(SIGCLD, SIG_IGN); /* ignore child death */
	(void)signal(SIGHUP, SIG_IGN); /* ignore terminal hangups */
	for(i=0;i<32;i++)
		(void)close(i);		/* close open files */
	(void)setpgrp();		/* break away from process group */
	logger(LOG,"nweb starting",argv[1],getpid());
	/* setup the network socket */
	if((listenfd = socket(AF_INET, SOCK_STREAM,0)) <0)
		logger(ERROR, "system call","socket",0);
	port = atoi(argv[1]);
	if(port < 0 || port >60000)
		logger(ERROR,"Invalid port number (try 1->60000)",argv[1],0);
	serv_addr.sin_family = AF_INET;
	serv_addr.sin_addr.s_addr = htonl(INADDR_ANY);
	serv_addr.sin_port = htons(port);
	if(bind(listenfd, (struct sockaddr *)&serv_addr,sizeof(serv_addr)) <0)
		logger(ERROR,"system call","bind",0);
	if( listen(listenfd,64) <0)
		logger(ERROR,"system call","listen",0);
	for(hit=1; ;hit++) {
		length = sizeof(cli_addr);
		if((socketfd = accept(listenfd, (struct sockaddr *)&cli_addr, &length)) < 0)
			logger(ERROR,"system call","accept",0);
		if((pid = fork()) < 0) {
			logger(ERROR,"system call","fork",0);
		}
		else {
			if(pid == 0) { 	/* child */
				(void)close(listenfd);
				web(socketfd,hit); /* never returns */
			} else { 	/* parent */
				(void)close(socketfd);
			}
		}
	}
}
